### ISSUE TITLE
Model score quickly becomes NaN even though nearly equivalent network works fine in Keras

### ISSUE THREAD

**Issue by wmeddie** [deeplearning4j/deeplearning4j#3750]
### Issue Description

While trying to create recreate a keras benchmark I found that our score calculation becomes NaN after a few batches while the almost equivalent keras model does not.  Regardless the network is able to train. However it does not get the same 99.25% accuracy after 12 epochs.

Sample code and output: https://gist.github.com/wmeddie/42b44e1a8e8427226c579cc852174b79

### Version Information

Deeplearning4j 0.8.1-SNAPSHOT (as of afternoon 7/27)
Windows 10
CUDA 8 CuDNN 6

**Comment by raver119** [comment]
What happens with nd4j-native backend?

**Comment by wmeddie** [comment]
Exactly the same.  I'll check to see what happens if I use DataType.Double.

**Comment by AlexDBlack** [comment]
@wmeddie there are some issues with current master, which *may* have impacted results you are seeing (see dev channel). I'd recommend trying again after this is confirmed to be resolved.

**Comment by wmeddie** [comment]
They go away on CUDA with that learning rate as well.  But it is only delaying the problem.  I leaning rate of 0.0001 only gets us to 0.87 ish accuracy.  Doing 2000 epochs instead of 20 with the MLP, NaNs start appearing at epoch 120 and after epoch 165 it's almost always NaN.

There may be a difference between the meaning of learning rate for Keras and DL4J, but 1.0 is the default there and it works fine.

**Comment by AlexDBlack** [comment]
As far as I can tell, the only significant difference in the Keras implemantion is the manual/explicit clipping of the softmax values. I have added clipping to our implementation here:
https://github.com/deeplearning4j/nd4j/pull/2030

After adding it, running the DL4J code given here https://gist.github.com/wmeddie/42b44e1a8e8427226c579cc852174b79 gives the following output:
https://gist.github.com/AlexDBlack/efbf4572f8ed810b5048d2766fe6666e

**Comment by AlexDBlack** [comment]
Fixed here: https://github.com/deeplearning4j/nd4j/pull/2030
